chapter/3/message-passing.md

```diff
@@ -1,11 +1,250 @@
+
+Actors in AmbientTalk/2 are representations of event loops. The message queue is the event queue, messages are events, asynchronous message sends are event notifications, and object methods are the event handlers. The event loop serially processes messages from the queue to avoid race conditions. Local objects within an actor are owned by that actor, which is the only entity allowed to directly execute methods on them. Like E, objects within an actor can communicate using synchronous or asynchronous methods of communication. Again similar to E, objects that are referenced outside of an actor can only be communicated to asynchronously by sending messages. Objects can additionally declare themselves serializable, which means they can be copied and sent to other actors for use as local objects. When this happens, there is no maintained relationship between the original object and its copy.
+
+AmbientTalk/2 uses the event loop model to enforce three essential concurrency control properties:
+
+* _Serial execution_: Events are processed sequentially from an event queue, so the handling of a single event is atomic with respect to other events.
+* _Non-blocking communication_: An event loop doesn't suspend computation to wait for other event loops, instead all communication happens strictly as asynchronous event notifications.
+* _Exclusive state access_: Event handlers (object methods) and their associated state belong to a single event loop, which has access to their mutable state. Mutation of other event loop state is only possible indirectly by passing an event notification asking for mutation to occur.
+
+The end result of all this decoupling and isolation of computation is that it is a natural fit for mobile ad hoc networks. In this domain, connections are volatile with limited range and transient failures. Removing coupling based on time or synchronization is a natural fit for the domain, and the communicating event-loop actor model is a natural model for programming these systems. AmbientTalk/2 provides additional features on top of the communicating event-loop model like service discovery. These enable ad hoc network creation as actors near each other can broadcast their existence and advertise common services that can be used for communication.
+
+AmbientTalk/2 is most notable as a reimagining of the communicating event-loops actor model for a modern use case. This again speaks to the broader advantages of actors and their applicability to solving the problems of distributed systems.
+
+# Active Objects
```

Comment:
Maybe this paper might help [*Mirror-based Reflection in AmbientTalk*](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.481.4964&rep=rep1&type=pdf) (below is a link):

http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.481.4964&rep=rep1&type=pdf

Here are some sections as comparison, that maybe might help:

*AmbientTalk/2 [10], an updated version of the language whose reflective API differs from the version presented in previous work [4] (pg. 5)*

*In previous work, we have discussed the metaobject protocol of AmbientTalk/1 – the predecessor of the AmbientTalk language described in this paper – to develop language features specifically for mobile ad hoc networks [4]. In AmbientTalk/1, an actor is represented as an active object which executes in a thread of its own, has a message queue and a dedicated behavior describing the methods that may be asynchronously invoked on the active object. This behavior object contains base-level application methods as well as metalevel methods used to hook into the actor’s metaobject protocol. Intercession is made possible by making the active object implement a metalevel method, which is only distinguishable from a base-level method by its reserved name. (pg. 34)*

*In AmbientTalk/1 reflection is neither stratified nor encapsulated: base-level code can be affected by the implementation details of metalevel code such as that of a new language feature. Because the base- and metalevels are not partitioned into separate namespaces, name clashes between the two levels can occur. For example, a base-level method may accidentally be regarded as a metalevel method simply because its name accidentally matches that of a metalevel operation. Finally, AmbientTalk/1 had no metaobject protocol for objects, only for actors. (pg. 34)*